src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs

```diff
@@ -773,38 +774,48 @@ public override BoundNode VisitArrayAccess(BoundArrayAccess node)
-                return base.VisitArrayAccess(node)!;
+                resultExpr = base.VisitArrayAccess(node)!;
             }
-
-            var indexType = VisitType(node.Indices[0].Type);
-            var F = _factory;
-
-            BoundNode resultExpr;
-            if (TypeSymbol.Equals(
-                indexType,
-                _compilation.GetWellKnownType(WellKnownType.System_Range),
-                TypeCompareKind.ConsiderEverything))
+            else
             {
-                // array[Range] is compiled to:
-                // System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray(array, Range)
-
-                Debug.Assert(node.Expression.Type is { TypeKind: TypeKind.Array });
-                var elementType = ((ArrayTypeSymbol)node.Expression.Type).ElementTypeWithAnnotations;
-
-                resultExpr = F.Call(
-                    receiver: null,
-                    F.WellKnownMethod(WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__GetSubArray_T)
-                        .Construct(ImmutableArray.Create(elementType)),
-                    ImmutableArray.Create(
-                        VisitExpression(node.Expression),
-                        VisitExpression(node.Indices[0])));
+                var indexType = VisitType(node.Indices[0].Type);
+                var F = _factory;
+
+                if (TypeSymbol.Equals(
+                    indexType,
+                    _compilation.GetWellKnownType(WellKnownType.System_Range),
+                    TypeCompareKind.ConsiderEverything))
+                {
+                    // array[Range] is compiled to:
+                    // System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray(array, Range)
+
+                    Debug.Assert(node.Expression.Type is { TypeKind: TypeKind.Array });
+                    var elementType = ((ArrayTypeSymbol)node.Expression.Type).ElementTypeWithAnnotations;
+
+                    resultExpr = F.Call(
+                        receiver: null,
+                        F.WellKnownMethod(WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__GetSubArray_T)
+                            .Construct(ImmutableArray.Create(elementType)),
+                        ImmutableArray.Create(
+                            VisitExpression(node.Expression),
+                            VisitExpression(node.Indices[0])));
+                }
+                else
+                {
+                    resultExpr = base.VisitArrayAccess(node)!;
+                }
             }
-            else
+
+            if (_inCompoundAssignmentReceiver &&
+                node.Expression.Type is ArrayTypeSymbol { ElementType: { TypeKind: TypeKind.TypeParameter, IsValueType: false } } &&
+                resultExpr is BoundArrayAccess arr)
             {
-                resultExpr = base.VisitArrayAccess(node)!;
+                resultExpr = arr.Update(arr.Expression, arr.Indices, inCompoundAssignmentReceiver: true, arr.Type);
```

Comment:
Do we only set `inCompoundAssignmentReceiver` when the element type is a type parameter? If so then I think we should rename that property such that it captures. Based on the current name I was confused looking at the changes to `EmitAddress.cs` as to why more baselines didn't change to have the `.readonly` prefix on them.